homeassistant/components/azure_servicebus/notify.py

```diff
@@ -0,0 +1,83 @@
+"""Support for azure service bus notification."""
+import json
+
+from azure.servicebus.aio import Message, ServiceBusClient
+import voluptuous as vol
+
+from homeassistant.components.notify import (
+    ATTR_DATA,
+    ATTR_TARGET,
+    ATTR_TITLE,
+    PLATFORM_SCHEMA,
+    BaseNotificationService,
+)
+from homeassistant.const import CONTENT_TYPE_JSON
+import homeassistant.helpers.config_validation as cv
+
+CONF_CONNECTION_STRING = "connection_string"
+CONF_QUEUE_NAME = "queue"
+CONF_TOPIC_NAME = "topic"
+
+ATTR_ASB_MESSAGE = "message"
+ATTR_ASB_TITLE = "title"
+ATTR_ASB_TARGET = "target"
+
+PLATFORM_SCHEMA = vol.All(
+    cv.has_at_least_one_key(CONF_QUEUE_NAME, CONF_TOPIC_NAME),
+    PLATFORM_SCHEMA.extend(
+        {
+            vol.Required(CONF_CONNECTION_STRING): cv.string,
+            vol.Exclusive(
+                CONF_QUEUE_NAME, "output", "Can only send to a queue or a topic."
+            ): cv.string,
+            vol.Exclusive(
+                CONF_TOPIC_NAME, "output", "Can only send to a queue or a topic."
+            ): cv.string,
+        }
+    ),
+)
+
+
+async def async_get_service(hass, config, discovery_info=None):
+    """Get the notification service."""
+    connection_string = config[CONF_CONNECTION_STRING]
+    queue_name = config.get(CONF_QUEUE_NAME)
+    topic_name = config.get(CONF_TOPIC_NAME)
+
+    servicebus = ServiceBusClient.from_connection_string(
+        connection_string, loop=hass.loop
+    )
+
+    if queue_name:
+        client = servicebus.get_queue(queue_name)
```

Comment:
I'm looking at the library source. It's very weird. Getting the queue seems to be using the requests library, ie sync I/O.

https://github.com/Azure/azure-sdk-for-python/blob/c648cc4d18d2052bb57a6d8259f56ffa40c94b39/sdk/servicebus/azure-servicebus/azure/servicebus/aio/async_client.py#L117-L119

While at the same time they are using asyncio functions that must be run within the event loop when creating the client:
https://github.com/Azure/azure-sdk-for-python/blob/c648cc4d18d2052bb57a6d8259f56ffa40c94b39/sdk/servicebus/azure-servicebus/azure/servicebus/aio/async_client.py#L58

Since we're passing the loop explicitly it should be safe to run this in sync context to avoid making sync I/O inside the event loop.

I'm not sure what they are thinking when creating the library.

Reply:
They upgraded the library with breaking changes a little while back, and it seems like they try to combine 3 different ways of using the clients to not break too much for anyone.
Looks like they perhaps wasn't careful enough while reusing some parts...